refactor: updating new part 0: Honeycomb Dependencies and How To Install (#20)
brown-ccv/honeycomb-docs: docs/quick_start.md

title: Quick Start
---

> 

Reviewer: This looks great!

title: Quick Start
---

Before we can get started with Honeycomb, we must make sure our machines have the required prerequisites for Honeycomb to run. 
### 0. Installing Prerequisites
There are some prerequisites that are required across all operating systems to run Honeycomb. For macOS users, we highly recommend using Homebrew to expedite installing these prerequisites.

The following are required for this project: 
1. git
2. Node.js
3. Electron
4. Firebase CLI

> #### Installing Prerequisites with Homebrew (for macOS)

Reviewer 2: ```suggestion
#### Installing Prerequisites on macOS (Homebrew)
```

##### Installing Homebrew 
Paste the following in a macOS Terminal and press enter to install Homebrew. 

  

>     /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"

Reviewer 2: ```suggestion
```
    /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```
```

 
Note: *When this command is run, additional installations may be automatically added by Homebrew based on your machine's configurations.
Homebrew will prompt you to `Press RETURN to continue or any other key to abort`. Please continue with these installations by pressing `Return`*

##### Running Brewfile 
A Brewfile is a script that developers can use to install multiple software tools that are required to run an application. 

In the project directory, locate the file named `Brewfile` and on the command line, run the following command:
    

>     brew bundle

Reviewer 2: Adding "```" around text will add a copy button! Plus something like "```js" will add text coloring and what not
```suggestion
```
    brew bundle
```
```

    
This will call the `brew` command on each of the packages listed, effectively downloading and installing all of the desired prerequisites to your project. 

### 1. Start your new task from our template repository

> To start a new task locally, in development mode follow these steps:

Reviewer 2: I think we can axe this line? The sentence below it says basically the same thing

Author: Oops my mistake

- Git also provides safety mechanisms to ensure collaborative work is done systematically and allow a large group of developers to work on one project together without getting in each others way. 

#### Node.js 
- Open-source JavaScript platform that allows users to build network applications quickly. It is just another way to execute code on your computer. 
- Node.js is a popular choice as it is cross-platform, so a Node.js app can run on all major operating systems. 
- Electron, one of the other main dependencies of the Honeycomb app, requires the Node.js runtime environment.

#### Electron 
- Framework for building desktop applications using JavaScript, HTML, and CSS 
- Requires you to have one JavaScript codebase that will create cross-platform apps that work on Windows, MacOS, and Linux.
- Honeycomb relies on Electron to package the cross-platform desktop application. 

##### Prerequisties Specific to Windows

- Visual Studio:
Install the latest version of [Visual Studio](https://visualstudio.microsoft.com/downloads/) with the Desktop Development for C++ Workflow. 

> To add the workflow, follow [these instructions](https://docs.microsoft.com/en-us/cpp/build/vscpp-step-0-installation?view=msvc-160#:~:text=If%20you%20have%20Visual%20Studio,Then%2C%20choose%20Modify).

Reviewer 2: I believe this is only a prerequisite for building local Windows applications? I wonder if we should leave it somewhere else in the docs that talks about building for Windows specifically, it's a note in section 5 right now

Reviewer 3: 1. It would be good to add a link to the "version control" doc page under the GIT section . 
2. I think you can get rid of the "Electron, one of the other main dependencies of the Honeycomb app, requires the Node.js runtime environment." bullet in the Node.js section. You and add in the Electron section that it requires the Node.js runtime environment if you want.

##### Prerequisties Specific to MacOS

- Command Line Tools: Type in the terminal

    ```

>     xcode-select --install

Reviewer 3: I believe homebrew will ask to install xcode when you first do the homebrew install with `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"` since it's a requirement for homebrew. So we could just explain that during the homebrew installation instead of having this extra step here.

Author: Updated!

Follow the guide on electron docs to [install and configure python and its modules](https://www.electronjs.org/docs/development/build-instructions-macos#python) 

##### Prerequisties Specific to Linux

- Python 3.7 or later (with support for TLS 1.2):
Install [Python](https://www.python.org/downloads/)

- Clang:
Install [Clang](https://clang.llvm.org/get_started.html) or follow installation instructions on the [electron docs](https://www.electronjs.org/docs/development/build-instructions-linux#prerequisites)

- Development headers of GTK 3 and libnotify:
Follow installation instructions on the [electron docs](https://www.electronjs.org/docs/development/build-instructions-linux#prerequisites)

**Note: Ubuntu 18.10 or later recommended (or another distro new enough to support GLIBC_2.28)**
**Note: restart computer after all installs are complete**

You can read more in-depth about Electron at [electron documentation](https://www.electronjs.org/docs/development/build-instructions-gn) for your specific OS.

#### How To Install - Brew
Note: This applies only to MacOS users. 
- Using Brew will help simplify the task of downloading the prerequisite software introduced above. 

> - To make this process easier we will be using a brew file which will complete the installation in one simple command line command.

Reviewer 2: I'm wondering if we should split this entire section into max/windows/linux instead of going back and forth?

Author: True, but do we have directions for npm install for windows/linux?

Reviewer 2: Yeah that's true. Windows we just have to instruct the user to run on a bash terminal instead of the default one but that doesn't need to be its own section

##### What is Brew (HomeBrew)?
- Brew is an open-source software built specifically for macOS that makes it easy to install software on your machine. 
- It is very straightforward and makes the installation of software very easy for users of all experience levels. 
- Installation happens with the `brew` command, which gives us access to thousands of command-line utilities. 

> ##### What is a brewfile?

Reviewer 3: Need Homebrew install directions before the "What is a brewfile" section.

Author: Thanks George! Updated

  
    /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
 

> Note: There may be additional instructions you have to follow, depending on user's machine configurations.

Reviewer 2: ```suggestion
Note: There may be additional instructions you have to follow, depending on your machine's configurations.
```

*Checkout the [Version Control](https://brown-ccv.github.io/honeycomb-docs/docs/version_control) page for more information about working with git*

### 3. Install NPM Packages
With Node.js installed in Step 0, we are now able to utilize the `npm` command in the terminal. 

npm is a command-line utility for Node.js that makes it easy for you to install packages and maintain them throughout the lifecycle of your application. 

> Honeycomb, like most applications, contains a template called `package.json`. This file contains three distinctive parts that we interact with using a series of npm commands.

Reviewer 2: Can these be combined into one paragraph instead of being on separate lines?

Author: Definitely. How i see it, we can combine the first two lines. I plan to put the npm install instruction right after this small paragraph. In my opinion the `Honeycomb, like most applications, contains ....` is the start of the subsection that breaks down the `package.json` so I think that should be the start of its own line. What do you think?

##### 1. Metadata
![](assets/packagelock1.png)
The metadata provides information about the project such as the name, author and description. 

##### 2. Dependencies
![](assets/packagelock2.png)

Dependencies are packages that the project rely on for it to function properly. 

They are formatted in specific key/value pairs where every key is a name of the package and the value is the version range that’s acceptable.

Please run the following command to trigger the installation of all the dependencies listed in the `package.json`.

>     npm install

Reviewer 2: I think this should go before `####1` - tell people to run `npm install` and if there interested in what's going on under the hood they can read on. Also, be sure to do 

\```
npm install
\```

*Checkout the [Version Control](https://brown-ccv.github.io/honeycomb-docs/docs/version_control) page for more information about working with git*

### 3. Install NPM Packages
With Node.js installed in Step 0, we are now able to utilize the `npm` command in the terminal.
`npm` is a command-line utility for Node.js that makes it easy for you to install packages and maintain them throughout the lifecycle of your application. 

Please run the following command to trigger the installation of all the dependencies relevant to this application.

```
npm install
```

Honeycomb, like most applications, contains a template called `package.json`. This file contains three distinctive parts that we interact with using a series of npm commands.

##### 1. Metadata

> ![](assets/packagelock1.png)

Reviewer: I think we might be best putting in the pure code here, rather than a screenshot.

```json
{
  "name": "honeycomb",
  "description": "all-in-one task starter app with jsPsych + React + Electron + psiturk ",
  "author": {
    "name": "Brown CCV",
    "email": "ccv-bot@brown.edu",
    "url": "ccv.brown.edu"
  },
```

The metadata provides information about the project such as the name, author and description. 

##### 2. Dependencies

> ![](assets/packagelock2.png)

Reviewer: Similarly here: 

```json
"dependencies": {
    "@brown-ccv/behavioral-task-trials": "^2.0.0",
    "@fortawesome/fontawesome-free": "^5.9.0",
    "bootstrap": "^5.2.0-beta1",
    "electron-log": "^4.4.8",
    "electron-squirrel-startup": "^1.0.0",
    "event-marker": "git+https://github.com/brown-ccv/event-marker.git",
}
```

Dependencies are packages that the project rely on for it to function properly. 

They are formatted in specific key/value pairs where every key is a name of the package and the value is the version range that’s acceptable.

Note: *If any changes are made to the dependencies section of the `package.json`, you must run `npm install` again to download the newly updated list of dependencies.*

##### 3. Scripts

> ![](assets/packagelock3.png)

Reviewer: And here
```json
"scripts": {
    "commit": "git-cz",
    "postinstall": "node version.js && npm run rebuild",
    "start": "cross-env BROWSER=\"none\" NODE_PATH=\"./src\" react-scripts start",
    "start:browser": "react-scripts start",
    "test": "react-scripts test",
    "prebuild": "electron-rebuild",
    "build": "react-scripts build",
}
```
